src/localplayer.cpp

```diff
@@ -319,46 +234,94 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 	bool touching_ground_was = touching_ground;
 	touching_ground = result.touching_ground;
 
+	// Max. distance (X, Z) over border for sneaking determined by collision box
+	// * 0.49 to keep the center just barely on the node
+	v3f sneak_max = m_collisionbox.getExtent() * 0.49;
+
+	if (m_sneak_ladder_detected) {
+		// restore legacy behaviour (this makes the m_speed.Y hack necessary)
+		sneak_max = v3f(0.4 * BS, 0, 0.4 * BS);
+	}
+
+	/*
+		If sneaking, keep on top of last walked node and don't fall off
+	*/
+	if (could_sneak && m_sneak_node_exists) {
+		const v3f sn_f = intToFloat(m_sneak_node, BS);
+		const v3f bmin = m_sneak_node_bb_top.MinEdge;
+		const v3f bmax = m_sneak_node_bb_top.MaxEdge;
+		const v3f old_pos = position;
+		const v3f old_speed = m_speed;
+
+		position.X = rangelim(position.X,
+			sn_f.X + bmin.X - sneak_max.X, sn_f.X + bmax.X + sneak_max.X);
+		position.Z = rangelim(position.Z,
+			sn_f.Z + bmin.Z - sneak_max.Z, sn_f.Z + bmax.Z + sneak_max.Z);
+
+		if (position.X != old_pos.X)
+			m_speed.X = 0;
+		if (position.Z != old_pos.Z)
+			m_speed.Z = 0;
+
+		if (m_speed.Y < 0) {
+			// Move player on top of the node when falling
+			// The m_speed.Y < 0 check reduces the climb speed on sneak ladders
+			position.Y = MYMAX(position.Y, sn_f.Y + bmax.Y);
+			if (position.Y > old_pos.Y)
+				m_speed.Y = 0;
+		}
+
+		if (m_speed.Y == 0)
+			touching_ground = true;
```

Comment:
Note for the reviewers:
This line allows jumping on node edges while sneaking. Without it, the player would be stuck as soon the sneak ladder isn't detected anymore -> when reaching its top.